File: DataStructuresAndAlgorithms/disjoint_sets/uva_p459.py
par=[0, 0, 2, 0, 2]      #parents A for ABD and C for CE

def find(u):                                                                    #basic disjoint format
    """
    Finds the representitive node within a subset.

    Parameters:
        u <int>: provides the index mapped to the element in question
    """
    if u==par[u]:
        return u
    par[u]=find(par[u])
    return par[u]

# def connect(u,v):                                                             #basic disjoint format                     
#     """
#     Helps connect multiple nodes to a certain subset.

#     Parameters:
#         u,v <int>: The arguments are nodes that must be connected to the same subset
#     """
#     u=find(u)
#     v=find(v)
#     if u==v:
#         print("already chuddy buddies")
#     else:
#         par[max(u,v)]=min(u,v)

def count_sub(args,principle_case):                                   #Problem specific function
    """
    Counts the correct maximal substrings from our inputs
    Parameters:
        - args(string): Input string to be converted to index values and checked for maximal substring matching
                        according to principle case.
        - kwargs:
            - principle_case: Representitive index converted from principle_case string input for 
                            matching against input substrings.        
    """
    for i in [ord(i)-65 for i in args]:
        new_rep=find(i)
        if new_rep==principle_case:
            continue
        else:
            return False
    return True


if __name__=="__main__":
    cases=int(input())
    _=input()
    for i in range(cases):
        principle_case=find(ord(input())-65)
        counter=0
        while True:
            case=input()
            if case=="":
                break
            if count_sub(case,principle_case=principle_case):
                counter+=1
        print(counter)

File: DataStructuresAndAlgorithms/disjoint_sets/uva_p459.cpp
#include <bits/stdc++.h>
using namespace std;

int par[5]={0,0,2,0,2};

int find(int u){
    if (u==par[u]){
        return u;
    par[u]=find(par[u]);
    return par[u];
    }
}


int count_sub(char* args, int principle_case){
    for(int i=0;i<args.length();i++){
        int new_rep=find(i);
        if (new_rep==principle_case){
            continue;
        }
        else{
            return 0;
        }
    }
    return 1;
}

int main(){
    char args[2];
    int cases;
    scanf("%d", &cases);
    char space;
    for (int i=0;i<cases;i++){
        char a;
        scanf("%c\n", &a);
        int principle_case=find(int(a)-65);
        int counter=0;
        while (true){
            char u,v;
            scanf("%c",&u);
            if (u==NULL){
                break;
            }
            scanf("%c\n",&v);
            args={u,v}
            if count_sub(char* args, principle_case=principle_case){
                counter++;
            }
            }
        printf("%d\n",counter)

        }
    }


